### DISCUSSION TITLE
Nested virtualization required?

### DISCUSSION THREAD

**Discussion by iverberk** [cozystack/cozystack#489] [General]
I'm trying to understand how kubevirt runs the tenant worker nodes. As far as I understand a combination of qemu and KVM is used on the worker node to start a VM 'process'. 

So, when I install CozyStack on a VM-based management cluster (meaning Talos runs on vms itself), does that mean that the vms (tenant worker nodes) created by kubevirt require nested virtualization to have the full speed benefits? I believe the worker nodes are launched on the management cluster?

**Comment by kvaps** [comment]
Hey, yeah, we're focusing on bare-metal servers and using KubeVirt to spawn VMs. This means Cozytack covers all infrastructure needs from the very lowest layers (hardware, OS, and hypervisor) up to the highest levels of applications and managed services.

![image](https://github.com/user-attachments/assets/c4dfda10-6ec3-47c2-88f4-c8b0655c9843)

**Comment by iverberk** [comment]
Thanks for replying. So, would you then agree that CozyStack is not as useful when running in a virtualized environment? For instance, unless you request metal type machines in the cloud, you'll always get a virtual machine. Same with vmware on-prem? I don't think kubevirt can fulfill it's role in that case and you would need to integrate with the vm-provider directly.

**Reply by kvaps** [reply]
If you are considering using Cozystack to run Kubernetes clusters and VMs, then running it inside the VMs is not particularly useful. However, if you plan to host other services like DBaaS, it might be worth considering.

There is also an opportunity to extend functionality. Cozystack can be enhanced to support virtualization using other providers for Cluster API like VMware, Proxmox, and others. While we have plans to include this feature, it is in our long-term roadmap.

If you are interested in adding this feature, you have two options:

1.  Implement it yourself: Please join our community meeting to discuss how to get started. (just add this events to your [Google Calendar](https://calendar.google.com/calendar?cid=ZTQzZDIxZTVjOWI0NWE5NWYyOGM1ZDY0OWMyY2IxZTFmNDMzZTJlNjUzYjU2ZGJiZGE3NGNhMzA2ZjBkMGY2OEBncm91cC5jYWxlbmRhci5nb29nbGUuY29t) or [iCal](https://calendar.google.com/calendar/ical/e43d21e5c9b45a95f28c5d649c2cb1e1f433e2e653b56dbbda74ca306f0d0f68%40group.calendar.google.com/public/basic.ics)) or [Telegram group](https://t.me/cozystack).
2.  Sponsor the feature development: You can schedule a meeting with our core team using the following link: [aenix.io/contact](https://aenix.io/contact/)

**Reply by iverberk** [reply]
Thanks again for your response. At this time I'm not interested in extending the functionality but it's good to know that it can be done. Good luck with the project, it's very interesting!

**Reply by kvaps** [reply]
Thank you @iverberk
